src/java/org/apache/cassandra/index/sai/IndexContext.java

```diff
@@ -0,0 +1,435 @@
+        this.indexType = Objects.requireNonNull(indexType);
+
+        this.config = config;
+        this.indexMetrics = new IndexMetrics(this);
+        this.validator = TypeUtil.cellValueType(columnMetadata, indexType);
+
+        this.primaryKeyFactory = PrimaryKey.factory(clusteringComparator);
+
+        // We currently only support the NoOpAnalyzer
+        this.indexAnalyzerFactory = AbstractAnalyzer.fromOptions(getValidator(), Collections.emptyMap());
+        this.queryAnalyzerFactory = AbstractAnalyzer.fromOptions(getValidator(), Collections.emptyMap());
+    }
+
+    public AbstractType<?> keyValidator()
+    {
+        return partitionKeyType;
+    }
+
+    public PrimaryKeyFactory keyFactory()
+    {
+        return primaryKeyFactory;
+    }
+
+    public ClusteringComparator comparator()
+    {
+        return clusteringComparator;
+    }
+
+    public String getKeyspace()
+    {
+        return keyspace;
+    }
+
+    public String getTable()
+    {
+        return table;
+    }
+
+    public long index(DecoratedKey key, Row row, Memtable mt)
+    {
+        MemtableIndex current = liveMemtables.get(mt);
+
+        // We expect the relevant IndexMemtable to be present most of the time, so only make the
+        // call to computeIfAbsent() if it's not. (see https://bugs.openjdk.java.net/browse/JDK-8161372)
+        MemtableIndex target = (current != null)
+                               ? current
+                               : liveMemtables.computeIfAbsent(mt, memtable -> new MemtableIndex(this));
+
+        long start = Clock.Global.nanoTime();
+
+        long bytes = 0;
+
+        if (isNonFrozenCollection())
+        {
+            Iterator<ByteBuffer> bufferIterator = getValuesOf(row, FBUtilities.nowInSeconds());
+            if (bufferIterator != null)
+            {
+                while (bufferIterator.hasNext())
+                {
+                    ByteBuffer value = bufferIterator.next();
+                    bytes += target.index(key, row.clustering(), value);
+                }
+            }
+        }
+        else
+        {
+            ByteBuffer value = getValueOf(key, row, FBUtilities.nowInSeconds());
+            target.index(key, row.clustering(), value);
+        }
+        indexMetrics.memtableIndexWriteLatency.update(Clock.Global.nanoTime() - start, TimeUnit.NANOSECONDS);
+        return bytes;
+    }
+
+    public void renewMemtable(Memtable renewed)
+    {
+        for (Memtable memtable : liveMemtables.keySet())
+        {
+            // remove every index but the one that corresponds to the post-truncate Memtable
+            if (renewed != memtable)
+            {
+                liveMemtables.remove(memtable);
+            }
+        }
+    }
+
+    public void discardMemtable(Memtable discarded)
+    {
+        liveMemtables.remove(discarded);
+    }
+
+    public RangeIterator searchMemtable(Expression e, AbstractBounds<PartitionPosition> keyRange)
```

Comment:
```suggestion
    public RangeIterator searchMemtables(Expression e, AbstractBounds<PartitionPosition> keyRange)
```

Reply:
I would prefer `searchMemtableIndexes` because this is what it is going to do. At the same time I'll change `liveMemtables` to `liveMemtableIndexMap`. Some of these names a too terse to be meaningful.